validator_client/src/doppelganger_service.rs

```diff
@@ -0,0 +1,159 @@
+                }
+            }
+        };
+
+        executor.spawn(interval_fut, "doppelganger_service");
+        Ok(())
+    }
+
+    async fn detect_doppelgangers(&self) -> Result<(), String> {
+        let log = self.context.log().clone();
+
+        let slot = self.slot_clock.now().ok_or("Unable to read slot clock")?;
+        let epoch = slot.epoch(E::slots_per_epoch());
+
+        // get all validators in the doppelganger detection epoch
+        let validator_map = self
+            .validator_store
+            .initialized_validators()
+            .read()
+            .get_doppelganger_detecting_validators_by_epoch(epoch);
+
+        for (detection_epoch, validators) in validator_map {
+            // This is to make sure we always check for attestations at the doppelganger detection epoch
+            // and all epochs after the current epoch. We want to avoid the current epoch so we don't
+            // pick up attestations from our own validator on restart.
+            let epochs: Vec<Epoch> =
+                (((detection_epoch - Epoch::new(DOPPELGANGER_DETECTION_EPOCHS)).as_u64() + 1)
+                    ..=detection_epoch.as_u64())
+                    .map(Epoch::new)
+                    .collect();
+
+            let epochs_slice = epochs.as_slice();
+            let validators_slice = validators.as_slice();
+            info!(log, "Monitoring for doppelgangers"; "epochs" => ?epochs);
+
+            let doppelganger_detected = self
+                .beacon_nodes
+                .first_success(RequireSynced::Yes, |beacon_node| async move {
+                    beacon_node
+                        .get_lighthouse_seen_validators(validators_slice, epochs_slice)
+                        .await
+                        .map_err(|e| format!("Failed query for seen validators: {:?}", e))
+                        .map(|result| result.data)
+                })
+                .await
+                .map_err(|e| format!("Failed query for seen validators: {}", e));
+
+            // Send shutdown signal if necessary
+            match doppelganger_detected {
+                Ok(doppelgangers) => {
+                    if !doppelgangers.is_empty() {
+                        crit!(
+                            log,
+                            "Doppelganger detected! Shutting down. Ensure you aren't already \
+                                         running a validator client with the same keys.";
+                                         "doppelganger_indices" => ?doppelgangers
+                        );
+
+                        let _ = self
+                            .context
+                            .executor
+                            .shutdown_sender()
+                            .try_send("Doppelganger detected.")
+                            .map_err(|e| format!("Could not send shutdown signal: {}", e))?;
+                    }
+                }
+                Err(e) => {
+                    warn!(
```

Comment:
```suggestion
                    crit!(
```

Might be nice to make this loud since it can prevent validators from running.